### PR TITLE
AP_OSD: Add fence indicator panel

### PR DESCRIPTION
**Pull request by Hwurzburg**
Adds Fence status icon panel to OSD....shows if enabled, disabled, or in breach....works even if in Reporting Only mode for breaches, if fence is enabled...
https://youtu.be/goUN2Mgka8A

### CODE REVIEW

**Comment by Hwurzburg** [comment]
other options for a fence icon:
Screenshot from 2021-04-30 13-13-49
Screenshot from 2021-04-30 13-14-40
Screenshot from 2021-04-30 13-15-55
Screenshot from 2021-04-30 13-17-54
Screenshot from 2021-04-30 13-20-16

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 1]
```diff
@@ -961,6 +962,22 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info[] = {
...
+    // @Param: FENCE_EN_X
+    // @DisplayName: FENCE_EN_X
+    // @Description: Horizontal position on screen
+    // @Range: 0 29
+
+    // @Param: FENCE_EN_Y
+    // @DisplayName: FENCE_EN_Y
+    // @Description: Vertical position on screen
+    // @Range: 0 15
+    AP_SUBGROUPINFO(fence, "FENCE", 59, AP_OSD_Screen, AP_OSD_Setting),
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_Screen.cpp:980]:
```suggestion
    // @Param: FENCE_EN_Y
    // @DisplayName: FENCE_EN_Y
    // @Description: Vertical position on screen
    // @Range: 0 15
    AP_SUBGROUPINFO(fence, "FENCE_EN", 59, AP_OSD_Screen, AP_OSD_Setting),
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
Found the problem.  It is actually relatively straight forward from the test output in CI - once you can catch the subtlety!

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 2]
```diff
@@ -961,6 +962,22 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info[] = {
...
     // @Range: 0 15
     AP_SUBGROUPINFO(restvolt, "RESTVOLT", 58, AP_OSD_Screen, AP_OSD_Setting),
+
+    // @Param: FENCE_EN
+    // @DisplayName: FENCE_EN
+    // @Description: Displays indication of fence enable and breach
+    // @Values: 0:Disabled,1:Enabled
+
+    // @Param: FENCE_EN_X
+    // @DisplayName: FENCE_EN_X
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_Screen.cpp:972]:
```suggestion
    // @Param: FENCE_X
    // @DisplayName: FENCE_X
```

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 3]
```diff
@@ -961,6 +962,22 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info[] = {
...
+    // @Description: Horizontal position on screen
+    // @Range: 0 29
+
+    // @Param: FENCE_EN_Y
+    // @DisplayName: FENCE_EN_Y
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_Screen.cpp:977]:
```suggestion
    // @Param: FENCE_Y
    // @DisplayName: FENCE_Y
```

**Review by peterbarker — COMMENTED** [review]
*Sigh*.  Maybe this time?

Its almost like we need to have `FENEN` rather than `FENCE` here, so

`FENEN_EN`, `FENEN_X`, `FENEN_Y` rather than `FENCE_EN`, `FENCE_X` and `FENCE_Y`

But that's a different argument.

**Comment by Hwurzburg** [comment]
@peterbarker if it works I will squash before devcall, which I wont make....on the road for 5 days

**Comment by Hwurzburg** [comment]
@peterbarker darn! still not emitting! this beast will not die!

File: libraries/AP_OSD/AP_OSD.h [hunk 4]
```diff
@@ -184,6 +184,7 @@ class AP_OSD_Screen : public AP_OSD_AbstractScreen
     AP_OSD_Setting callsign{false, 0, 0};
     AP_OSD_Setting vtx_power{false, 0, 0};
     AP_OSD_Setting hgt_abvterr{true, 23, 7};
+    AP_OSD_Setting fence{true,14,9};
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD.h:187]:
```suggestion
    AP_OSD_Setting fence{true, 14, 9};
```

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 5]
```diff
@@ -961,6 +962,22 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info[] = {
...
+    // @Param: FENCE_X
+    // @DisplayName: FENCE_X
+    // @Description: Horizontal position on screen
+    // @Range: 0 29
+
+    // @Param: FENCE_Y
+    // @DisplayName: FENCE_Y
+    // @Description: Vertical position on screen
+    // @Range: 0 15
+    AP_SUBGROUPINFO(fence, "FENCE_EN", 59, AP_OSD_Screen, AP_OSD_Setting),
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_Screen.cpp:980]:
```suggestion
    AP_SUBGROUPINFO(fence, "FENCE", 59, AP_OSD_Screen, AP_OSD_Setting),
```

I think this was my first wild-assed solution - which was wrong and needs to be reverted.

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 6]
```diff
@@ -1948,6 +1967,21 @@ void AP_OSD_Screen::draw_hgt_abvterr(uint8_t x, uint8_t y)
...
 
+
+void AP_OSD_Screen::draw_fence(uint8_t x, uint8_t y)
+{
+    AC_Fence *fenceptr = AP::fence();
+    if (fenceptr == nullptr) {
+       return;
+    }
+    if (fenceptr->enabled()) {
+    backend->write(x, y, fenceptr->get_breaches(), "%c", SYM_FENCE_ENABLED);
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_Screen.cpp:1978]:
```suggestion
        backend->write(x, y, fenceptr->get_breaches(), "%c", SYM_FENCE_ENABLED);
```

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 7]
```diff
@@ -1948,6 +1967,21 @@ void AP_OSD_Screen::draw_hgt_abvterr(uint8_t x, uint8_t y)
...
+    } else {
+    backend->write(x, y, false, "%c", SYM_FENCE_DISABLED);
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_Screen.cpp:1980]:
```suggestion
        backend->write(x, y, false, "%c", SYM_FENCE_DISABLED);
```

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Comment by Hwurzburg** [comment]
@peterbarker so line 187,the macro, has a requirement for spaces in the arguments? that was the culprit?
in the dev-call, once everyone decides on the icon, I will change and squash and then someone merge after the CI passes again, if agreed..

**Comment by tridge** [comment]
I like the diagonal FEN myself, I think it is clearer

**Review by peterbarker — CHANGES_REQUESTED** [review]
DevCall outcome was a request to move from the little fence picture to the diagonal `FEN` picture.

**Review by peterbarker — APPROVED** [review]
LGTM.

The travis failure can be ignored.

**Review by tridge — APPROVED** [review]

**Pull request merged**